### PR TITLE
Rust language support + gitignore practice

### PR DESCRIPTION
**Pull request by TheEdward162**
It's a me-a, Edwardio

## Description

This PR adds support for the Rust programming language. It includes a language and component detector and a package inspector. It also includes a "correctly set gitignore" practice for Rust.

## Motivation and Context

Avoiding school.

Rust is a great programming language and this PR provides a base for practices like the [clippy](https://github.com/rust-lang/rust-clippy) linter and [rustfmt](https://github.com/rust-lang/rustfmt) formatter to be implemented in the future.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] I have updated the documentation accordingly (not sure about this one).
- [x] I have added new practice to practice list in README.md (I added the language as WIP).
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/inspectors/package/RustPackageInspector.ts [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+        } else {
+          this.cargoManifest = {
+            package: RustPackageInspector.parsePackage(cargoManifestToml['package']),
+            ...RustPackageInspector.parseDependencySet(cargoManifestToml),
+            target: RustPackageInspector.parseTarget(cargoManifestToml['target']),
+            bin: RustPackageInspector.parseBin(cargoManifestToml['bin']),
+            profile: cargoManifestToml['profile'],
+          };
+
+          this.packages = RustPackageInspector.extractPackages(this.cargoManifest, this.cargoLock);
```
Comment (adelkahomolova) [inline src/inspectors/package/RustPackageInspector.ts:76]: Keep the convetion and name the method the same as in all the Package inspectors. The code will be much more readable.
```suggestion
          this.packages = RustPackageInspector.addPackages(this.cargoManifest, this.cargoLock);
```

File: src/inspectors/package/RustPackageInspector.ts [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      result ?? {
+        value: '0.0.0',
+        major: '0',
+        minor: '0',
+        patch: '0',
+      }
+    );
+  }
+
+  private static extractPackages(manifest: CargoManifest, lockFile?: CargoLock): Package[] {
```
Comment (adelkahomolova) [inline src/inspectors/package/RustPackageInspector.ts:314]:
```suggestion
  private static addPackages(manifest: CargoManifest, lockFile?: CargoLock): Package[] {
```

File: src/inspectors/package/RustPackageInspector.ts [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+  path: string;
+}
+
+interface DependecySet {
+  dependencies: ReadonlyArray<ManifestDependency>;
+  'dev-dependencies': ReadonlyArray<ManifestDependency>;
+  'build-dependencies': ReadonlyArray<ManifestDependency>;
+}
+// https://doc.rust-lang.org/cargo/reference/manifest.html
+export interface CargoManifest extends DependecySet {
```
Comment (adelkahomolova) [inline src/inspectors/package/RustPackageInspector.ts:372]: I don't think it's necessary to have all the optional properties. All the packages in all the languages we support return the same object. Please, return also the same object. Keep our conventions.

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import { RustGitignoreCorrectlySetPractice } from './RustGitignoreCorrectlySetPractice';
+import { PracticeEvaluationResult } from '../../model';
+import { TestContainerContext, createTestContainer } from '../../inversify.config';
+
+const GITIGNORE_MISSING = `
```
Comment (adelkahomolova) [inline src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts:5]: It's not necessary. To check badly set `.gitignore` you can do just the `'...'`

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+# foo
+foo/
+
+bar/*.baz
+`;
+
+const GITIGNORE_PRESENT = `
```
Comment (adelkahomolova) [inline src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts:12]: It's not the variable. Use our convention and name it `basicGitignore`

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  private IGNORE_CHECKS: [RegExp, string][] = [
+    [/target\//, 'target/'],
+    [/\*\*\/\*\.rs\.bk/, '**/*.rs.bk'],
+  ];
+
+  async isApplicable(ctx: PracticeContext): Promise<boolean> {
+    return ctx.projectComponent.language === ProgrammingLanguage.Rust;
+  }
+
+  private static async parseGitignore(fileInspector: IFileInspector): Promise<{ raw: string; rules: string[] }> {
```
Comment (adelkahomolova) [inline src/practices/Rust/RustGitignoreCorrectlySetPractice.ts:29]: Please, keep the same structure as we have in other practices for checking correctly set gitignore. E.g. C#, JavaScript, ... It's quite hard to check it if all the code is correct.

**Review by adelkahomolova — CHANGES_REQUESTED** [review]
Hi, @TheEdward162 , I really appreciate the time you invest in our repository. The pull request is really nice and working. But I have one huge request. Please, keep the conventions that are already there in the project. I know your code looks nice but it's extremely difficult to oriented in and to check it for me since there is support for 5 languages and all of them have the detectors, inspectors and their practices have the same structure. I have to look at the line by line to understand if your code is working the same way I would expect. 
Also, you use `.then` instead of `await/async`, keyword `function` instead of arrow functions. I left you many comments. Please, refactor your code. It would be really awesome if the dx-scanner could help also Rust developers. I hope I didn't disappoint you with my comment, but we want to keep the quality and readability of code as high as possible. I also think that it can move you forward in your programming skills trying to keep already set conventions. I totally understand that you can have different ones, but open-source is not just about adding working code, but keeping the set convention in every project is a must. Thank you for your time. :)

**Comment by adelkahomolova** [comment]
Also, the binding of `RustPackageInspector` is missing. See the [languageContextBinding](https://github.com/DXHeroes/dx-scanner/blob/master/src/contexts/language/languageContextBinding.ts#L73)

**Comment by TheEdward162** [comment]
> Also, the binding of `RustPackageInspector` is missing. See the [languageContextBinding](https://github.com/DXHeroes/dx-scanner/blob/master/src/contexts/language/languageContextBinding.ts#L73)

https://github.com/DXHeroes/dx-scanner/blob/b6915355dc58050967b1dba9e7beeb4ef75b6a3f/src/contexts/language/languageContextBinding.ts#L75

I ran the scanner on a real repo (local) repo and it worked, is there something apart from this that I should bind?

File: src/detectors/Rust/RustComponentDetector.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    // * `src/bin/*.rs`
+    // * `src/bin/*/main.rs`
+    // * a `bin` entry in the manifest
+    const hasBinaryFile = Promise.all([
+      this.fileInspector.exists('src/main.rs'),
+      this.fileInspector
+        .readDirectory('src/bin')
+        .then(async (entries) => {
+          for (const entry of entries) {
+            const entryPath = `src/bin/${entry}`;
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:29]: Yes, according to the official Cargo documentation on [Package Layout](https://doc.rust-lang.org/cargo/guide/project-layout.html) this covers all possible instances when Cargo generates an executable binary.

**Comment by adelkahomolova** [comment]
> > Also, the binding of `RustPackageInspector` is missing. See the [languageContextBinding](https://github.com/DXHeroes/dx-scanner/blob/master/src/contexts/language/languageContextBinding.ts?rgh-link-date=2020-10-29T15%3A56%3A51Z#L73)
> 
> https://github.com/DXHeroes/dx-scanner/blob/b6915355dc58050967b1dba9e7beeb4ef75b6a3f/src/contexts/language/languageContextBinding.ts#L75
> 
> I ran the scanner on a real (local) repo and it worked, is there something apart from this that I should bind?

Oh, sorry. I missed that out. My bad 🙈

File: src/detectors/Rust/RustComponentDetector.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    // Cargo outputs a binary output when it finds:
+    // * `src/main.rs`
+    // * `src/bin/*.rs`
+    // * `src/bin/*/main.rs`
+    // * a `bin` entry in the manifest
+    const hasBinaryFile = Promise.all([
+      this.fileInspector.exists('src/main.rs'),
+      this.fileInspector
+        .readDirectory('src/bin')
+        .then(async (entries) => {
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:27]: I honestly believe that the functional style programming pattern here is more readable, as it better demonstrates the flow of information. With try/catch I'd either have to include too much code into the block or use a `let` binding which we rust people hate (joke).

As for the `Promise.all` call, I would really rather want some kind of a `Promise.raceWhile` to short-circuit on the first rejection or the first resolution with the value of `true`.

Anyhow, the alternative code would look like this:
```
const hasMain = await this.fileInspector.exists('src/main.rs');
let hasSrcBin = false;
try {
    const binFolderEntries = await this.fileInspector.readDirectory('src/bin');
    for (const entry of binFolderEntries) {
      const entryPath = `src/bin/${entry}`;
      if (
        (entry.endsWith('.rs') && (await this.fileInspector.isFile(entryPath))) ||
        (await this.fileInspector.isFile(`${entryPath}/main.rs`))
      ) {
        hasSrcBin = true;
      }
    }
} catch (e) {
    if (e.code === 'ENOENT') {
        // ignore
    } else {
        throw e;
    }
}
const hasManifestBin = _.get(this.packageInspector.cargoManifest, 'bin.length', 0) > 0;
const hasBinaryFile = hasMain || hasSrcBin || hasManifestBin;
```

I personally don't like the monstrous try/catch and the fact that adding another condition would mean having to add the condition calculation and then updating the `hasBinaryFile` which is now all handled by the `Promise.all([]).then(a => a.some(<identity>)` function and I don't even have to come up with additional variable names :D

Either way, if you insist on strictly using imperative `async/await` with `try/catch` I'll do the second variant.

File: src/detectors/Rust/RustComponentDetector.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        framework: ProjectComponentFramework.UNKNOWN,
+        language: langAtPath.language,
+        path: langAtPath.path,
+        platform: ProjectComponentPlatform.BackEnd,
+        repositoryPath: undefined,
+        type: ProjectComponentType.Library,
+      });
+    }
+
+    if (await hasBinaryFile) {
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:58]: Strictly speaking you may have a standalone `Cargo.toml` file without any source files present or one with some rust sources of which neither generates any output. It is not a valid rust package, however, and Cargo will fail when attempting to build such a package. I'm not sure what to do in that case. Can you advise?

File: src/detectors/Rust/RustLanguageDetector.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+        return [
+          {
+            language: ProgrammingLanguage.Rust,
+            path: commonPath,
+          },
+        ];
+      });
+
+    const manifestFiles = await manifestFilesPromise;
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustLanguageDetector.ts:55]: In an ideal world the promises might be executed on another thread that does IO while the main thread continues execution until it truly needs the result (in this case it prepares all the conditions in promises and then resolves them when it actually needs them).

Whether or not that is actually happening depends on the implementation (i.e. node) and whether the promises are lazy or not, but I believe this pattern _may_ have positive performance impact.

I would also like to come back to your previous questions about `async/await` vs `.then`. When using `await` directly in the top-level function (in this case the `detectLanguage`) the whole function stalls while the value is resolved. That means each step that is now represented by `then` would be a separate top-level await and would prevent the current function from advancing, thus the IO operations to never happen in parallel.

When using `then` (or `await` nested in `then` callbacks) it is actually the internal promise resolution that is stalled, thus allowing the top-level function to continue executing (and preparing other conditions) until it needs the results. This allows (in theory, and depending on how node is interally implemented, maybe even in practice) a performance boost because the IO operations actually run in parallel.

Of course I don't have any benchmarks to back this up so if you'd like me to rewrite it I will, but I am confident in what I said.

File: src/inspectors/package/RustPackageInspector.spec.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  afterEach(() => {
+    containerCtx.virtualFileSystemService.clearFileSystem();
+  });
+
+  it('inits and loads packages', async () => {
+    containerCtx.virtualFileSystemService.setFileSystem({});
+
+    await inspector.init();
+
+    expect(inspector.hasPackageManagement()).toBe(false);
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/inspectors/package/RustPackageInspector.spec.ts:30]: You are correct. I have changed the name of the test to `inits correctly even with empty fs` as it actually tests that the package inspector correctly initializes and doesn't crash when it doesn't find `Cargo.toml`.

File: src/inspectors/package/RustPackageInspector.ts [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+  return typeof v === 'boolean' || v === undefined;
+};
+const isOptStringArray = function (v: unknown): v is string[] | undefined {
+  return v === undefined || (Array.isArray(v) && v.every((e) => typeof e === 'string'));
+};
+
+/**
+ * Similar to `lodash.conformsTo` but runs the check on `undefined` instead of failing.
+ */
+const conformsOptional = function (check: Record<string, (v: unknown) => boolean>, value: Record<string, unknown>): boolean {
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/inspectors/package/RustPackageInspector.ts:28]: Is it possible at all to add this to the linter? It did warn me about `function <name>()` and told me to use `const <name> = function ()` but I'd appreciate if it could then tell me to use arrow functions as well.

[the remark this answers is not in the document]
Comment (adelkahomolova) [reply src/inspectors/package/RustPackageInspector.ts:28]: I'm not sure actually. I can check it.

File: src/detectors/Rust/RustComponentDetector.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (adelkahomolova) [reply src/detectors/Rust/RustComponentDetector.ts:58]: I think it's quite the edge case that somebody would want to check a project without any source files present. We assume that people will check working projects with the dx-scanner. So my thought was to simplify code to just on if. `if (hasBinaryFile) { save the component info  - is library } else { the component is application }`. What do you think about it?

File: src/inspectors/package/RustPackageInspector.ts (hunk 3 above)
Comment (TheEdward162) [reply src/inspectors/package/RustPackageInspector.ts:372]: I'm not sure what you mean. The inspector doesn't actually return anything, instead it sets the inherited variable `packages` and fills it with the dependencies and the code does do this.

I did fill out all the properties according to the official Cargo documentation because it makes it easier for future implementors to build off of it if they'd like to detect properties like the `edition` or `license`.

However, I do realize that I have exposed the `cargoManifest` variable publicly on the inspector. I did this because I needed access to the properties of cargo manifest for the sake of the compoment detector and I believe practice checkers will need access to this file as well, as they would for example for checking rust edition or license as I mentioned above. I wasn't sure how to do it better as there is no precedence because all the existing practices either work with files outside the package management or with packages themselves. The checks I'm talking about are akin to looking for the `license` key in `package.json` or simillar, which isn't done anywhere in the existing codebase yet.

File: src/detectors/Rust/RustComponentDetector.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:58]: I agree with returning something in case nothing is detected. However there is also the question of multiple return values. From the array return type I have assumed that the compoment detector may return multiple compoments for each configurations it finds.

One Cargo package can contain at most one library and 0 or more binaries, all in the same package, with the same lock file and same dependencies. This led me to believe that I should expose this by returning 2 different components from this detector. What is your opinion on that?

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts (hunk 4 above)
Comment (TheEdward162) [reply src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts:5]: I'm actually also checking that it preserves the old value [here](https://github.com/DXHeroes/dx-scanner/blob/b6915355dc58050967b1dba9e7beeb4ef75b6a3f/src/practices/Rust/RustGitignoreCorrectlySetPractice.spec.ts#L79) while writing the new one.

I do this because I found out that when the component detector detects more than one component at the same path (which happens when there is a cargo package that compiles to both an executable and a library at the same path) then the fix handler runs for both components, so it runs twice. Since it runs asynchronously it actually has a **race condition** and appends the fix twice.

Thus I was forced to implement the fix using `writeFile` and overwrite the whole gitignore file so I don't append to it twice. Note that it still actually has a race condition and the only real solution would be to implement some kind of a locking mechanism for fixers to synchronize over.

File: src/inspectors/package/RustPackageInspector.ts [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+    this.fileInspector = fileInspector;
+  }
+
+  async init(): Promise<void> {
+    try {
+      this.debug('RustPackageInspector init started');
+
+      const cargoLockString = await this.fileInspector
+        .readFile('Cargo.lock')
+        .catch((err) => (err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err)));
```
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/inspectors/package/RustPackageInspector.ts:50]: It doesn't actually have different files with packages, I'm simply reading the lock file so I can correctly fill out the `lockfileVersion` in `Package` type.

In my opinion the question here comes down to whether we are willing to tolerate the additional IO operation when calling `.exists`. The pattern you are suggesting comes down to calling the `lstat` and `readdir` POSIX api functions back to back. The current solution only calls `readdir` and reacts to its error. I thus believe that the current solution is more performant. However, I do agree about readability suffering somewhat. 

I dare suggest two alternatives from which even the `JavaPackageInspector` might benefit in readability:

1. Define the following adaptor on `IFileInspector` (and implement it on `FileInspector`):
```js
static allowNotFound<T>(p: Promise<T>): Promise<T | undefined> {
    return p.catch((err) => (err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err)))
}
```
and call it like `FileInspector.allowNotFound(fileInspector.readFile(<args>))`.

2. On the same interface, define more lenient function for `readFile` and `readDirectory` in pattern of:
```js
readFileOptional(path: string): Promise<string | undefined> {
    this.readFile(path).catch((err) => err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err))
}
```
and then check for `undefined`.

Whether or not you are willing to accept this is entirely up to you of course, but I do believe it is a better fix for this missing functionality.

I'd also like to point out that I didn't come up with this on my own, I simply saw it in the [implementation of](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/inspectors/FileInspector.ts#L114) `scanFor` function because I originally wanted to use it instead of `readDirectory` in the component detector.

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.ts (hunk 6 above)
Comment (TheEdward162) [reply src/practices/Rust/RustGitignoreCorrectlySetPractice.ts:29]: I think this could warrant some kind of a base class that each language can extend for the gitignore practice specifically (and also other practices that need to be customizable to some degree as well).

For example, I can imagine something like this
```ts
abstract class GitignorePresentPracticeBase extends PracticeBase {
  protected constructor(
    protected readonly applicableLanguages: ProgrammingLanguage[],
    protected readonly ruleChecks: { check: RegExp, fix: string }[] 
  ) {
    super();
  }

  protected static async parseGitignore(fileInspector: IFileInspector): Promise<{ raw: string; rules: string[] }> {
    const raw = await fileInspector.readFile('.gitignore');
    const rules = raw.split('/\r?\n/').filter((line) => !line.startsWith('#') && line.trim() !== '');

    return {
      raw,
      rules,
    };
  }
  
  async isApplicable(ctx: PracticeContext): Promise<boolean> {
    return this.applicableLanguages.includes(ctx.projectComponent.language);
  }

  async evaluate(ctx: PracticeContext): Promise<PracticeEvaluationResult> {
    if (!ctx.root.fileInspector) {
      return PracticeEvaluationResult.unknown;
    }

    const gitignore = await GitignorePresentPracticeBase.parseGitignore(ctx.root.fileInspector);
    const allFound = this.ruleChecks.every((rule) => gitignore.rules.find((v) => rule.check.test(v)) !== undefined);

    if (allFound) {
      return PracticeEvaluationResult.practicing;
    }

    return PracticeEvaluationResult.notPracticing;
  }

  async fix(ctx: FixerContext) {
    const inspector = ctx.fileInspector?.basePath ? ctx.fileInspector : ctx.root.fileInspector;
    if (!inspector) return;

    const gitignore = await GitignorePresentPracticeBase.parseGitignore(inspector);
    const fixes = this.ruleChecks.filter((rule) => gitignore.rules.find((v) => rule.check.test(v)) === undefined).map(
      (rule) => rule.fix,
    );

    if (fixes.length > 0) {
      let fixesString = fixes.join('\n');
      if (fixesString.length > 0) {
        fixesString = '\n# added by `dx-scanner --fix`\n' + fixesString;
      }
      // TODO: Fix the race condition when two components are at the same path
      await inspector.writeFile('.gitignore', `${gitignore.raw}${fixesString}`);
    }
  }
}
```

then the rust practice checker would be much slimmer. In fact, it would factor a big chunk of common code out of `RubyGitignoreCorrectlySetPractice`, `JsGitignoreCorrectlySetPractice`, `CSharpGitignoreCorrectlySetPractice` and a moderate chunk out of `TsGitignoreCorrectlySetPractice` and `JavaGitignoreCorrectlySetPractice`.

```ts
@DxPractice({
  id: 'Rust.GitignoreCorrectlySet',
  name: 'Set .gitignore correctly',
  impact: PracticeImpact.high,
  suggestion: 'Set patterns in the .gitignore file as usual.',
  reportOnlyOnce: true,
  url: 'https://github.com/github/gitignore/blob/master/Rust.gitignore',
  dependsOn: { practicing: ['LanguageIndependent.GitignoreIsPresent'] },
})
export class RustGitignoreCorrectlySetPractice extends GitignorePresentPracticeBase {
  constructor() {
    super(
      [ProgrammingLanguage.Rust],
      [
        { check: /target\//, fix: 'target/' },
        { check: /\*\*\/\*\.rs\.bk/, fix: '**/*.rs.bk'}
      ]
    );
  }

  async evaluate(ctx: PracticeContext): Promise<PracticeEvaluationResult> {
    const result = await super.evaluate(ctx);

    if (result === PracticeEvaluationResult.notPracticing) {
      this.data.details = [
        {
          type: ReportDetailType.text,
          text: 'You should ignore the `target` folder, and for libraries also the `Cargo.lock` file.',
        },
      ];
    }
    return result;
  }
}
```

reads better, doesn't it?

Please consider. If you'd be open to this change, I'll open a separate PR which would also update all existing gitignore practices to use such base class.

**Comment by TheEdward162** [comment]
It looks like the CI from the last commit failed because it hit some kind of a rate limit on fetching the github repo in a runtime test. Not sure what I should do about it.

File: src/detectors/Rust/RustComponentDetector.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jakub-Vacek) [reply src/detectors/Rust/RustComponentDetector.ts:27]: Hi👋 Nice job 👍 Please stick to the `async/await`. Main reason is to avoid `Promise.all` (we already have some problems with it). Also you probably can break out of the `for` loop after you set `hasSrc = true`. Or you can use `array.some()`

[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:27]: Okay, I'll change it. Just out of curiosity, may I know what your problems with `Promise.all` are?

I can't really use `array.some()` because I do `await` inside the check loop but the break is a fair point. 👍

**Comment by adelkahomolova** [comment]
> It looks like the CI from the last commit failed because it hit some kind of a rate limit on fetching the github repo in a runtime test. Not sure what I should do about it.

Hi, it's just the GitHub API Rate limit because for some practices, we call the GH API and it has quite a small rate limit for unauthorized users. Nothing to worry about. I'll restart the job in Travis.

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.ts (hunk 6 above)
Comment (Jakub-Vacek) [reply src/practices/Rust/RustGitignoreCorrectlySetPractice.ts:29]: This is interesting idea. Problem can be with constructor. `GitignorePresentPracticeBase`has to be, from IOC point of view, just another practice. It means the it can access it's dependencies via `ctx: PracticeContext`. Also don't forget to call `this.setData` before `return PracticeEvaluationResult.notPracticing;`.

Comment (TheEdward162) [reply src/practices/Rust/RustGitignoreCorrectlySetPractice.ts:29]: `GitignorePresentPracticeBase` is abstract, same as `PracticeBase`, I'm not introducing any new concepts. The constructor doesn't matter because it's all static data in there anyway. It's just there to canonically pass information to the base class so that the implementations don't differ when they all do the same. Please realize that all the gitignore practices do exatly the same parsing and regex matching, and each of them repeats the same code. Only the Java and Typescript practices do some additional work afterwards, but they still go through `parseGitignore -> matchRegexes -> checkIfTheyAllMatch` and honestly they even do it pretty inefficiently at that. Having it in a common class should allow easier refactoring and better performance tweaking.

Additional;y, `this.setData` is actually not defined on `PracticeBase`. I see that is it an ad-hoc convention, but even between all the practices it's not consistent as sometimes [it is private,](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/CSharp/C%23GitignoreCorrectlySetPractice.ts#L83) sometimes [public and takes arguments](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/Java/JavaDependenciesVersionMajorLevel.ts#L58) and sometimes [it is inline](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/LanguageIndependent/CorrectCommitMessagesPractice.ts#L43).

That is exactly why in my comment I proposed to override `evaluate` method, call `super.evaluate()` and then set the data as I need it depending on the result. I think this is the exact problem inheritance is supposed to solve.

File: src/detectors/Rust/RustComponentDetector.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (adelkahomolova) [reply src/detectors/Rust/RustComponentDetector.ts:27]: The problem was, that when we used the `Promise.all()` in the other project with more async methods inside (each returns a different return type) we had to unforce the types because it was translated to `undefined` (see [this issue](https://github.com/microsoft/TypeScript/issues/33752)). When we add more than 10 methods inside the `Promise.all()` it crashed. There was an unreasonable cap number of methods we can pass inside the `Promise.all()`.
To sum it up: 
1) There was a problem with types.
2) There was a problem with the number of passed methods in `Promise.all()`.
Maybe there were more, but I don't remember it. Maybe @Jakub-Vacek ?

File: src/practices/Rust/RustGitignoreCorrectlySetPractice.ts (hunk 6 above)
Comment (Jakub-Vacek) [reply src/practices/Rust/RustGitignoreCorrectlySetPractice.ts:29]: Good point 👍 I am aware of repeated code that's why I like your idea. 🙂

Yes, it is an ad-hoc convention. 

Calling `setData` in child class is OK. My point was just not to forget calling it.

You can open separate PR for this change 🙂
I really like your approach 🤩

File: src/detectors/Rust/RustComponentDetector.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (TheEdward162) [reply src/detectors/Rust/RustComponentDetector.ts:27]: Hmm I see, It is defined in the typescript library with overloads for up to `T10`
```ts
all<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(values: readonly [T1 | PromiseLike<T1>, T2 | PromiseLike<T2>, T3 | PromiseLike<T3>, T4 | PromiseLike<T4>, T5 | PromiseLike<T5>, T6 | PromiseLike<T6>, T7 | PromiseLike<T7>, T8 | PromiseLike<T8>, T9 | PromiseLike<T9>, T10 | PromiseLike<T10>]): Promise<[T1, T2, T3, T4, T5, T6, T7, T8, T9, T10]>;
```
In my opinion it is not unreasonable, it is simply arbitrary because there are no variadic generics in typescript (yet! But we did get variadic tuple generics in TS4). For example, Rust does the same thing for tuples up to some limit because it also doesn't have variadic generics. I think it's better than nothing.

However I'd also like to refute this: This only applies to when you pass a tuple of more than 10 promises and at least two of them have different return types. Then instead of one of the typed overloads you'll invoke the `all<T>(values: readonly (T | PromiseLike<T>)[]): Promise<T[]>;` overload which requires that each promise has type `T` so you'll have to define `T` as an union of all the types appearing in the tuple. This is of course a problem when you rely on typing like this, for example in the patterns like `const [foo, bar] = Promise.all([fooPromise, barPromise]);`.

However, my solution does not. All promises passed into the `Promise.all` call return `boolean`, which is actually also apparent by the subsequent call to `.then((a) => a.some((i) => i))` (identity function => checking for truthyness). I'll of course still respect your preference with `async/await` but I wanted to make sure it is clear that this is not the same case as you ran into :)

File: src/inspectors/package/RustPackageInspector.ts (hunk 3 above)
Comment (adelkahomolova) [reply src/inspectors/package/RustPackageInspector.ts:372]: Yes, but we apply [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), meaning a programmer should not add functionality until deemed necessary. And adding the `license`, `edition`, or other is not necessary for now. So if you keep the same structure as other inspectors - do you still need the `cargoManifest` public?

Comment (TheEdward162) [reply src/inspectors/package/RustPackageInspector.ts:372]: Yes, I need access to it in the component detector. Regardless of how it turns out with the 2 components return value issue I still need it to check for the existence of a `bin` entry in the manifest.

**Review by prokopsimek — APPROVED** [review]
✅❤️
